### PR TITLE
Playwright: Mitigate Timeouts

### PR DESCRIPTION
**Pull request by ardelato**
## Description

If we are running all the Playwright tests locally then the tests will often timeout. There are two main reasons this occurs.

Currently, we have no set limit to the number of workers we will use locally. Therefore, we could potentially have 4-8 workers running in parallel, all depending on the number of cores on the host machine. In CI, we use all available workers we can and since the GitHub action machine only has two cores, we really only use 2 workers in CI. To reduce the load on local runs, I've changed the number of workers to a quarter of available cores.

The other reason for the timeout is each worker will create its own custom Nextjs server but each server will use the same dist directory (i.e. `.next`). Therefore, while one instance is building files, another instance may be overwriting these files. As a result of this overlap, some files may not be present when trying to access a page which will cause a timeout. To resolve this, each worker will build to a different directory than the default `.next`.

### CR

In order to override the default dist folder for each custom server we need to set an environment variable because we cannot simply import the config file and override the `distDir` property. The `next.config.js` file is a regular Node.js module and not a JSON file.[^1] Then by using Playwright's `worker` context[^2] we can create a unique dist folder for each worker.

Lastly, this does not completely resolve timeouts. We do have some flaky tests that may fail so if we do want to run the entire Playwright test-suite then we should increment the number of retries. 

### QA

- [ ] Confirm running tests locally will not timeout
	```shell
	pnpm playwright:run --project="Desktop Chrome" --retries=2
	```

Closes: #1323 

[^1]: [next.config.js](https://nextjs.org/docs/api-reference/next.config.js/introduction#:~:text=next.config.js%20is%20a%20regular%20Node.js%20module%2C%20not%20a%20JSON%20file.%20It%20gets%20used%20by%20the%20Next.js%20server%20and%20build%20phases%2C%20and%20it%27s%20not%20included%20in%20the%20browser%20build.)
[^2]: [Playwright Worker Info](https://playwright.dev/docs/api/class-workerinfo#worker-info-worker-index)

### CODE REVIEW

**Review by andyg0808 — APPROVED** [review]
CR :+1:

**Review by andyg0808 — APPROVED** [review]
CR :+1:

**Comment by deltuh-vee** [comment]
The collections_scroll test is still timing out, but the fix_kit test is now passing.
CleanShot 2023-02-06 at 14 28 05

**Comment by deltuh-vee** [comment]
QA 🎬 
.next folders are created for individual workers and are deleted after tests finish.

File: frontend/tests/playwright/test-fixtures.ts [hunk 1]
```diff
@@ -35,9 +36,14 @@ export const test = base.extend<
...
       // eslint-disable-next-line no-empty-pattern
-      async ({}, use) => {
+      async ({}, use, workerInfo) => {
+         process.env.NEXT_DIST_DIR = process.env.CI
+            ? './.next'
+            : `./.next-worker-${workerInfo.workerIndex}`;
          const props: CustomNextjsServer = await Server();
          await use(props);
+         // Clean up the dist directory after the worker is done.
+         exec('rm -rf .next-worker-*');
```
Comment (danielbeardsley) [inline frontend/tests/playwright/test-fixtures.ts:46]: I would think we'd want / need an await here as this is an async process.

Comment (ardelato) [reply frontend/tests/playwright/test-fixtures.ts:46]: Although `exec` is an `async` process, we are not doing anything with the return value of `exec` nor are we running dependent logic after it. Therefore, the `await` would have no effect.

Await Not Needed

Comment (andyg0808) [reply frontend/tests/playwright/test-fixtures.ts:46]: I'd double-check whether the test runner could potentially quit before this finishes running, or whether it'll block properly for this. Could change the command to a `sleep` possibly to just force it to wait.

Comment (danielbeardsley) [reply frontend/tests/playwright/test-fixtures.ts:46]: Andrew captured what I was referring to but it seems like we're safe here as node will wait on outstanding invocations (of `exec` at least):
image

**Pull request merged**